test/taskrun_test.go

```diff
@@ -208,3 +248,112 @@ spec:
 		t.Fatalf("-got, +want: %v", d)
 	}
 }
+
+func TestTaskRunModification(t *testing.T) {
+	taskrunModificationTest(t, false)
+}
+
+func TestWithSpireTaskRunModification(t *testing.T) {
+	taskrunModificationTest(t, true)
+}
+
+func taskrunModificationTest(t *testing.T, spireEnabled bool) {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	c, namespace := setup(ctx, t)
+	t.Parallel()
+
+	knativetest.CleanupOnInterrupt(func() { tearDown(ctx, t, c, namespace) }, t.Logf)
+	defer tearDown(ctx, t, c, namespace)
+
+	if spireEnabled {
+		originalConfigMapData := enableSpireConfigMap(ctx, c, t)
+		defer resetConfigMap(ctx, t, c, systemNamespace, config.GetFeatureFlagsConfigName(), originalConfigMapData)
+	}
+
+	taskRunName := "non-falsifiable-provenance"
+
+	t.Logf("Creating Task and TaskRun in namespace %s", namespace)
+	task := parse.MustParseTask(t, fmt.Sprintf(`
+metadata:
+  name: non-falsifiable
+  namespace: %s
+spec:
+  steps:
+  - image: ubuntu
+    script: |
+      #!/usr/bin/env bash
+      sleep 20
+      printf "hello" > "$(results.foo.path)"
+       printf "world" > "$(results.bar.path)"
+    results:
+    - name: foo
+    - name: bar
+`, namespace))
+	if _, err := c.TaskClient.Create(ctx, task, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Task: %s", err)
+	}
+	taskRun := parse.MustParseTaskRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  taskRef:
+    name: non-falsifiable
+`, taskRunName, namespace))
+	if _, err := c.TaskRunClient.Create(ctx, taskRun, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create TaskRun: %s", err)
+	}
+
+	t.Logf("Waiting for TaskRun in namespace %s to be in running state", namespace)
+	if err := WaitForTaskRunState(ctx, c, taskRunName, Running(taskRunName), "TaskRunRunning"); err != nil {
+		t.Errorf("Error waiting for TaskRun to start running: %s", err)
+	}
+
+	patches := []jsonpatch.JsonPatchOperation{{
+		Operation: "replace",
+		Path:      "/status/taskSpec/steps/0/image",
+		Value:     "not-ubuntu",
+	}}
+	patchBytes, err := json.Marshal(patches)
+	if err != nil {
+		t.Fatalf("failed to marshal patch bytes in order to stop")
+	}
+	t.Logf("Patching TaskRun %s in namespace %s mid run for spire to catch the un-authorized changed", taskRunName, namespace)
+	if _, err := c.TaskRunClient.Patch(ctx, taskRunName, types.JSONPatchType, patchBytes, metav1.PatchOptions{}, "status"); err != nil {
```

Comment:
may help to wait for `WaitForTaskRunState` to be `Running` so that we don't get overwritten by the initial `NotStarted` check in the controller.

Reply:
Line 310 above is `WaitForTaskRunState` to be `Running` before doing the patch.